src/goto-symex/path_storage.h

```diff
@@ -115,7 +115,31 @@ class path_storaget
   /// therefore may be referred to by a pointer.
   incremental_dirtyt dirty;
 
+  /// Generates a loop analysis for the instructions in goto_programt and
+  /// keys it against function ID.
+  void add_function_loops(const irep_idt &identifier, const goto_programt &body)
+  {
+    auto loop_iter = loop_analysis.find(identifier);
+    if(loop_iter == loop_analysis.end())
+    {
+      loop_analysis[identifier] =
+        std::make_shared<lexical_loopst>(lexical_loopst(body));
+    }
+  }
+
+  inline std::shared_ptr<lexical_loopst>
+  get_loop_analysis(const irep_idt &function_id)
+  {
+    INVARIANT(
+      loop_analysis.find(function_id) != loop_analysis.end(),
+      "Attempted to find loop analysis that didn't exist. Need to call "
+      "add_function_loops before attempting a find.");
+    return loop_analysis.at(function_id);
+  }
+
 private:
+  std::unordered_map<irep_idt, std::shared_ptr<lexical_loopst>> loop_analysis;
```

Comment:
Why `shared_ptr`? Where are the other owners of the object?

Reply:
While `path_storage` is pretty central you absolutely should be able to hold loop information outside of its lifetime, it's not like the data is going to change at all. All I wanted is something that acts like a reference in managed languages, and this fit the bill well enough.